theano/scan_module/scan_opt.py

```diff
@@ -612,8 +612,11 @@ def add_requirements(self, fgraph):
         fgraph.attach_feature(gof.toolbox.ReplaceValidate())
 
     def apply(self, fgraph):
+        # Don't perform the optimization on as_while scans since it's
+        # impossible to know for how many steps it will run.
         nodelist = [x for x in fgraph.toposort()
-                    if isinstance(x.op, scan_op.Scan)]
+                    if (isinstance(x.op, scan_op.Scan) and
+                        not x.op.as_while]
```

Comment:
This need a test. Seem ok for now, but I don't agree we can never push output of scan in while mode. After the scan, we know how many iteration was done. But we shouldn't push outputs used for by the condition. But we can let this for later.